storage/csv/datasource.go

```diff
@@ -145,66 +320,86 @@ func (rs *RecordStream) initializeColumnsWithHeaderRow() error {
 	return nil
 }
 
-func (rs *RecordStream) initializeColumnsWithoutHeaderRow() (*execution.Record, error) {
-	row, err := rs.r.Read()
+func (rs *RecordStream) initializeColumnsWithoutHeaderRow() (map[octosql.VariableName]octosql.Value, error) {
+	line, err := rs.r.Read()
 	if err != nil {
 		return nil, errors.Wrap(err, "couldn't read row")
 	}
 
-	rs.aliasedFields = make([]octosql.VariableName, len(row))
-	for i := range row {
+	rs.aliasedFields = make([]octosql.VariableName, len(line))
+	for i := range line {
 		rs.aliasedFields[i] = octosql.NewVariableName(fmt.Sprintf("%s.col%d", rs.alias, i+1))
 	}
 
-	return execution.NewRecordFromSlice(rs.aliasedFields, parseDataTypes(row)), nil
+	aliasedRecord := make(map[octosql.VariableName]octosql.Value)
+	for i, v := range line {
+		aliasedRecord[rs.aliasedFields[i]] = execution.ParseType(v)
+	}
+
+	return aliasedRecord, nil
 }
 
-func (rs *RecordStream) Next(ctx context.Context) (*execution.Record, error) {
-	if rs.isDone {
-		return nil, execution.ErrEndOfStream
+func parseDataTypes(row []string) []octosql.Value {
+	resultRow := make([]octosql.Value, len(row))
+	for i, v := range row {
+		resultRow[i] = execution.ParseType(v)
 	}
 
-	if rs.first {
-		rs.first = false
+	return resultRow
+}
 
-		if rs.hasColumnHeader {
-			err := rs.initializeColumnsWithHeaderRow()
-			if err != nil {
-				return nil, errors.Wrap(err, "couldn't initialize columns for record stream")
-			}
+var offsetPrefix = []byte("csv_offset")
 
-			return rs.Next(ctx)
-		} else {
-			record, err := rs.initializeColumnsWithoutHeaderRow()
-			if err != nil {
-				return nil, errors.Wrap(err, "couldn't initialize columns for record stream")
-			}
+func (rs *RecordStream) loadOffset(tx storage.StateTransaction) error {
+	offsetState := storage.NewValueState(tx.WithPrefix(offsetPrefix))
 
-			return record, nil
-		}
+	var offset octosql.Value
+	err := offsetState.Get(&offset)
+	if err == storage.ErrNotFound {
+		offset = octosql.MakeInt(0)
+	} else if err != nil {
+		return errors.Wrap(err, "couldn't load csv offset from state storage")
 	}
 
-	line, err := rs.r.Read()
-	if err == io.EOF {
-		rs.isDone = true
-		rs.file.Close()
-		return nil, execution.ErrEndOfStream
-	}
+	rs.offset = offset.AsInt()
+
+	return nil
+}
 
+func (rs *RecordStream) saveOffset(tx storage.StateTransaction, curBatchSize int) error {
+	offsetState := storage.NewValueState(tx.WithPrefix(offsetPrefix))
+
+	rs.offset = rs.offset + curBatchSize
```

Comment:
I'm not sure if saveOffset should include this addition. That's pretty unexpected. Maybe move it out of here and then call saveOffset after adding the current batch size.

Reply:
moved `rs.offset = rs.offset + curBatchSize` before `saveOffset` call